Add Wasm testing in Chromium, not just command-line `v8`

directhex opened the pull request:
This should give us a better idea of any cases where browser and JS engine behaviour diverge.

directhex committed:
Initial target to put Chromium into correlation payload for WASM on Windows

directhex committed:
Merge remote-tracking branch 'origin/master' into add_browser_2_the_browsening

directhex committed:
Revert "Initial target to put Chromium into correlation payload for WASM on Windows"

This reverts commit 95706be6a3c46c836628906cbc212c3e8077595c.

directhex committed:
Save intermediate state

directhex committed:
Bump to xharness version with `test wasm-browser` support

directhex committed:
Fix running wasm tests in a browser

directhex committed:
Bump xharness again, for selenium fix

directhex committed:
Specify browser scenario for wasm testing

directhex committed:
Merge remote-tracking branch 'origin/master' into add_browser_2_the_browsening

directhex commented:
Failures are because https://dnceng.visualstudio.com/internal/_git/dotnet-helix-machines/pullrequest/9677 hasn't landed in prod Helix instance yet.

premun commented on eng/pipelines/runtime.yml:
```diff
@@ -334,6 +334,11 @@ jobs:
         testRunNamePrefixSuffix: Mono_$(_BuildConfig)
+        scenarios:
+        # Only one scenario per build for jobs which install xharness CLI
+        # until we fix https://github.com/dotnet/arcade/issues/5919
```
https://github.com/dotnet/arcade/issues/5919 should be fixed now, but you might need to update the Helix SDK since Arcade wasn't released yet

directhex replied:
I tried bumping Arcade as part of this for testing and EVERYTHING EXPLODED so I'll wait for a bump to hit master I think

premun replied:
@directhex bump Helix SDK, not arcade. Check the iOS draft PR and cherry pick that

directhex committed:
Merge remote-tracking branch 'origin/master' into add_browser_2_the_browsening

directhex committed:
Bump Arcade to fix Helix bug

directhex committed:
Revert "Bump Arcade to fix Helix bug"

This reverts commit 10243dccf7707cd007176d3416f4fdf4bb2a7134.

directhex committed:
Try bumping Helix SDK again

directhex committed:
Merge remote-tracking branch 'origin/master' into add_browser_2_the_browsening

directhex committed:
Bump xharness, for supportFiles in Chromium fix

directhex committed:
Merge remote-tracking branch 'origin/master' into add_browser_2_the_browsening

directhex committed:
Bump xharness, for timeout fix

directhex committed:
Merge remote-tracking branch 'origin/master' into add_browser_2_the_browsening

directhex committed:
Merge remote-tracking branch 'origin/master' into add_browser_2_the_browsening

akoeplinger commented on eng/testing/tests.mobile.targets:
```diff
@@ -14,7 +14,7 @@
   <PropertyGroup Condition="'$(TargetOS)' == 'Browser'">
     <!-- We need to set this in order to get extensibility on xunit category traits and other arguments we pass down to xunit via MSBuild properties -->
-    <RunScriptCommand>$HARNESS_RUNNER wasm test --engine=$(JSEngine) $(JSEngineArgs) --js-file=runtime.js -v --output-directory=$XHARNESS_OUT -- --run WasmTestRunner.dll $(AssemblyName).dll</RunScriptCommand>
+    <RunScriptCommand>$HARNESS_RUNNER wasm test$TESTINBROWSER --app=. --engine=$(JSEngine) $(JSEngineArgs) --js-file=runtime.js -v --output-directory=$XHARNESS_OUT -- --run WasmTestRunner.dll $(AssemblyName).dll</RunScriptCommand>
```
nit: I'd prefer substituting the whole command rather than appending only the `-browser` portion.

E.g.
```suggestion
    <RunScriptCommand>$HARNESS_RUNNER wasm $XHARNESS_COMMAND --app=. --engine=$(JSEngine) $(JSEngineArgs) --js-file=runtime.js -v --output-directory=$XHARNESS_OUT -- --run WasmTestRunner.dll $(AssemblyName).dll</RunScriptCommand>
```

and then setting `XHARNESS_COMMAND` to `test` or `test-browser` depending on Scenario.

directhex committed:
Merge remote-tracking branch 'origin/master' into add_browser_2_the_browsening

directhex committed:
Wholesale swap-out of test command for wasm, not null-or-suffix fiddling

steveisok commented:
Definitely looks like we're still blocked by https://github.com/dotnet/runtime/issues/41160 and https://github.com/dotnet/runtime/issues/41269.

radical committed:
Merge remote-tracking branch 'origin/master' into add_browser_2_the_browsening

radical committed:
[wasm][tests] Use websocket to write pipe console messages to the test

.. runner.

radical commented:
Corresponding `xharness` PR - https://github.com/dotnet/xharness/pull/309 .

steveisok commented:
@directhex Please bump xharness and try again when the xharness PR lands.

directhex commented:
@steveisok yep, got my eye on it

directhex committed:
Merge remote-tracking branch 'origin/master' into add_browser_2_the_browsening

directhex committed:
Try bumping xharness

directhex committed:
Merge branch 'add_browser_2_the_browsening' of github.com:directhex/runtime into add_browser_2_the_browsening

directhex committed:
Bump xharness in global.json too

directhex committed:
--app is not optional

radical commented on src/mono/wasm/runtime-test.js:
```diff
@@ -18,9 +41,6 @@ if (is_browser) {
 	}
 }
 
-if (is_browser || typeof print === "undefined")
```
Um.. should this be:
```
if (typeof print === undefined)
   print = console.log;
```

.. for the non-browser case?

radical commented:
`System.Xml.XmlSchemaSet`:

```
[18:25:38] info: test-browser[0]
      Tests run: 585, Errors: 0, Failures: 0, Skipped: 0. Time: 10.16845s
      
[18:25:38] dbug: test-browser[0]
      Failed trying to read log messages via selenium: System.ArgumentException: http://127.0.0.1:38247/index.html?arg=--run&arg=WasmTestRunner.dll&arg=System.Xml.XmlSchemaSet.Tests.dll&arg=-notrait&arg=category%3dIgnoreForCI&arg=-notrait&arg=category%3dOuterLoop&arg=-notrait&arg=category%3dfailing - Access to fetch at 'http://www.w3.org/2001/03/xml.xsd' from origin 'http://127.0.0.1:38247' has been blocked by CORS policy: No 'Access-Control-Allow-Origin' header is present on the requested resource. If an opaque response serves your needs, set the request's mode to 'no-cors' to fetch the resource with CORS disabled.
         at Microsoft.DotNet.XHarness.CLI.Commands.Wasm.WasmBrowserTestRunner.RunSeleniumLogMessagePump(IWebDriver driver, CancellationToken token) in /_/src/Microsoft.DotNet.XHarness.CLI/Commands/WASM/Browser/WasmBrowserTestRunner.cs:line 222

< /snip>

XHarness exit code:
XHarness artifacts: ./xharness-output
```

Exit code is timeout.

directhex commented:
Wow, what an absolutely *phenomenal* improvement on the last run. Bravo, @radical

directhex commented:
```
[18:25:02] dbug: test-browser[0]
      [FAIL] System.Xml.Tests.AsyncReaderLateInitTests.ReadAfterInitializationWithUriOnAsyncReaderTrows
      
[18:25:02] dbug: test-browser[0]
      Assert.Throws() Failure
      
[18:25:02] dbug: test-browser[0]
      Expected: typeof(System.Net.Http.HttpRequestException)
      
[18:25:02] dbug: test-browser[0]
      Actual:   typeof(System.PlatformNotSupportedException): Cannot wait on monitors on this runtime.
      
[18:25:02] dbug: test-browser[0]
      ---- System.PlatformNotSupportedException : Cannot wait on monitors on this runtime.
```

radical commented:
`System.Xml`:

```
[18:25:02] dbug: test-browser[0]
      [FAIL] System.Xml.Tests.AsyncReaderLateInitTests.ReadAsyncAfterInitializationWithUriThrows
      System.InvalidOperationException : An asynchronous operation is already in progress.     
         at System.Xml.XmlAsyncCheckReader.CheckAsync()
         at System.Xml.XmlAsyncCheckReader.Dispose(Boolean disposing)
         at System.Xml.XmlReader.Dispose()
         at System.Xml.Tests.AsyncReaderLateInitTests.ReadAsyncAfterInitializationWithUriThrows()
         at System.Reflection.RuntimeMethodInfo.Invoke(Object obj, BindingFlags invokeAttr, Binder binder, Object[] parameters, CultureInfo culture)
```

```
[18:25:02] dbug: test-browser[0]
      [FAIL] System.Xml.Tests.AsyncReaderLateInitTests.ReadAfterInitializationWithUriOnAsyncReaderTrows

      Assert.Throws() Failure
      Expected: typeof(System.Net.Http.HttpRequestException)
      Actual:   typeof(System.PlatformNotSupportedException): Cannot wait on monitors on this runtime.
     
      ---- System.PlatformNotSupportedException : Cannot wait on monitors on this runtime.
         at System.Threading.Monitor.ObjWait(Int32 millisecondsTimeout, Object obj)
         at System.Threading.Monitor.Wait(Object obj, Int32 millisecondsTimeout)
         at System.Threading.ManualResetEventSlim.Wait(Int32 millisecondsTimeout, CancellationToken cancellationToken)
         at System.Threading.Tasks.Task.SpinThenBlockingWait(Int32 millisecondsTimeout, CancellationToken cancellationToken)
         at System.Threading.Tasks.Task.InternalWaitCore(Int32 millisecondsTimeout, CancellationToken cancellationToken)
         at System.Threading.Tasks.Task.InternalWait(Int32 millisecondsTimeout, CancellationToken cancellationToken)
         at System.Xml.XmlTextReaderImpl.FinishInitUriString()
         at System.Xml.XmlTextReaderImpl.FinishInit()
         at System.Xml.XmlTextReaderImpl.Read()
         at System.Xml.XmlAsyncCheckReader.Read()
         at System.Xml.Tests.AsyncReaderLateInitTests.<>c__DisplayClass8_0.<ReadAfterInitializationWithUriOnAsyncReaderTrows>b__0()

      ----- Inner Stack Trace -----
         at System.Threading.Monitor.ObjWait(Int32 millisecondsTimeout, Object obj)
         at System.Threading.Monitor.Wait(Object obj, Int32 millisecondsTimeout)
         at System.Threading.ManualResetEventSlim.Wait(Int32 millisecondsTimeout, CancellationToken cancellationToken)
         at System.Threading.Tasks.Task.SpinThenBlockingWait(Int32 millisecondsTimeout, CancellationToken cancellationToken)
         at System.Threading.Tasks.Task.InternalWaitCore(Int32 millisecondsTimeout, CancellationToken cancellationToken)
         at System.Threading.Tasks.Task.InternalWait(Int32 millisecondsTimeout, CancellationToken cancellationToken)
         at System.Xml.XmlTextReaderImpl.FinishInitUriString()
         at System.Xml.XmlTextReaderImpl.FinishInit()
         at System.Xml.XmlTextReaderImpl.Read()
         at System.Xml.XmlAsyncCheckReader.Read()
         at System.Xml.Tests.AsyncReaderLateInitTests.<>c__DisplayClass8_0.<ReadAfterInitializationWithUriOnAsyncReaderTrows>b__0()
      ```

directhex committed:
Patch System.Xml test which can't run on wasm

directhex committed:
xharness in correct file

directhex committed:
Don't try multi-scenario for iOS

radical commented:
I'm working on the failures in `XmlSchemaTest`

radical committed:
[wasm][debugger][tests] runtime-test.js - redirect console.{warn,info} also

radical commented:
fix for the failures seen in `XmlSchemaSet` - https://github.com/dotnet/xharness/pull/311 , which were failing the step even though all the tests had passed.

radical committed:
[wasm][tests] System.Net.Http FunctionalTests: skip the ones depending

.. on X509 certificates.
Tests discovery fails on
`System.Net.Http.Functional.Tests.HttpClientEKUTest..cctor()` which ends
up throwing PNSE for

`Certificates cctor threw System.PlatformNotSupportedException: System.Security.Cryptography.X509Certificates is not supported on this platform.`

Because it's a static ctor, `SkipOnMono` will not be helpful to skip
this one. So, we use `#if TARGETS_BROWSER` instead.

radical commented:
Should I open a separate PR for fixing `System.Net.Http/tests/FunctionalTests`? Or add it here?  https://gist.github.com/radical/f7f1a99ac9f4bd33554056401fdfeb4f

steveisok commented:
I'd probably just add it here.

radical committed:
[tests] System.Net.Http LoopbackServer: fix NRE

- This showed up when running System.Net.Http's FunctionalTests, as
creating a listening socket fails there with

`System.PlatformNotSupportedException : System.Net.Sockets is not supported on this platform.`

- Also, surface the exception so it can correctly fail tests.

Commit:
Added environment variable to help platform detect when we're running in the browser vs v8

Commit:
Merge branch 'add_browser_2_the_browsening' of github.com:directhex/runtime into add_browser_2_the_browsening

radical committed:
[wasm][tests] Don't include HttpClientEKUTest.cs for browser tests

radical committed:
[wasm][tests] Disable a hanging System.Net.Http test

System.Net.Http.Functional.Tests.SyncHttpHandler_HttpClient.HandlerTest.GetAsync_InvalidUrl_ExpectedExceptionThrown

radical committed:
[wasm][tests] Disable System.Net.Http tests hanging with the browser

`SocketsHttpHandler_HttpClientHandlerTest`

radical commented:
`System.Net.Http`'s `FunctionalTests` can now be enabled by changing `SkipOnMono` in `src/libraries/System.Net.Http/tests/FunctionalTests/AssemblyInfo.cs`.

And https://github.com/dotnet/xharness/pull/313 - has some nice additions.

lewing commented on src/mono/wasm/runtime-test.js:
```diff
@@ -5,9 +5,34 @@
 
 if (is_browser) {
 	// We expect to be run by tests/runtime/run.js which passes in the arguments using http parameters
+	window.real_print = console.log;
```
I want this in asap but after it lands I think we should consider sending all this as json over the websocket and expand it on the xharness side, we should be able to clean up the console output considerably.

pseudo code like
```
JSON.stringify ({ 'method': method, args: {...args}}) 
```
where method is { 'print', 'console.debug', ... }

and after that it might make sense to add a custom log provider so that send even more structure through.

lewing commented on src/mono/wasm/runtime-test.js:
```diff
@@ -5,9 +5,34 @@
+
+	consoleWebSocket = new WebSocket(consoleUrl);
+	consoleWebSocket.onopen = function(event) {
+		consoleWebSocket.send("browser: Console websocket connected.");
```
This might be a better place to force everything through print?  I sometimes use runtime-tests.js outside of xharness

radical replied:
Yeah, that makes sense. And I was thinking maybe the WebSocket thing should be optional? I found it useful to run the tests without the harness, directly in a browser.

lewing replied:
I think it is worth a follow-up once we have everything working in CI.

radical committed:
Bump xharness to get the new fixes

Update from channel - .NET Eng - Latest

Updating 'Microsoft.DotNet.XHarness.CLI': '1.0.0-prerelease.20474.1' => '1.0.0-prerelease.20476.2' (from build '20200926.2' of 'https://github.com/dotnet/xharness')
Updating 'Microsoft.DotNet.XHarness.TestRunners.Xunit': '1.0.0-prerelease.20474.1' => '1.0.0-prerelease.20476.2' (from build '20200926.2' of 'https://github.com/dotnet/xharness')

radical commented:
1 failing test remaining:

```
      [FAIL] System.Xml.Tests.AsyncReaderLateInitTests.ReadAsyncAfterInitializationWithUriThrows

      System.InvalidOperationException : An asynchronous operation is already in progress.     
         at System.Xml.XmlAsyncCheckReader.CheckAsync()
         at System.Xml.XmlAsyncCheckReader.Dispose(Boolean disposing)
         at System.Xml.XmlReader.Dispose()
         at System.Xml.Tests.AsyncReaderLateInitTests.ReadAsyncAfterInitializationWithUriThrows()
         at System.Reflection.RuntimeMethodInfo.Invoke(Object obj, BindingFlags invokeAttr, Binder binder, Object[] parameters, CultureInfo culture)
```

steveisok commented:
@radical I'm ok with skipping the last failing test.  

@akoeplinger Can you give this a final review to move things forward?

lewing committed:
Skip ReadAsyncAfterInitializationWithUriThrows without threading for now too

lewing commented:
@steveisok can we review System.Xml.Tests.AsyncReaderLateInitTests in the for the actual browser case after this.

akoeplinger committed:
Merge branch 'master' into add_browser_2_the_browsening

akoeplinger committed:
Rethrow exception in LoopbackServer

akoeplinger commented on src/libraries/System.Net.Http/tests/FunctionalTests/SocketsHttpHandlerTest.cs:
```diff
@@ -941,6 +943,7 @@ public sealed class SocketsHttpHandler_SchSendAuxRecordHttpTest : SchSendAuxReco
         public SocketsHttpHandler_SchSendAuxRecordHttpTest(ITestOutputHelper output) : base(output) { }
     }
 
+    [SkipOnMono("Tests hang with chrome. To be investigated", TestPlatforms.Browser)]
```
do we have a tracking issue for this?

lewing replied:
Should any of the SocketsHttpHandler tests be working?

marek-safar replied:
No, SocketsHttpHandler should throw PNSE on wasm-browser

akoeplinger replied:
Ok, let's follow up on this.

lewing commented:
Opened https://github.com/dotnet/runtime/issues/42852 to discuss the loopback server problem
